server/election.go

```diff
@@ -0,0 +1,284 @@
+		}
+		s.RunElection()
+
+		// if no leader was elected, wait 3 seconds then run another election
+		if s.leaderID == NO_LEADER {
+			s.logger.Info("No leader elected, waiting 3 seconds before trying again...")
+			time.Sleep(3 * time.Second)
+		}
+	}
+	return &LeaderResponse{Id: s.leaderID}, nil
+}
+
+// Checks if leader is alive every 1 second. If no response for 3 seconds, new election is held.
+func (s *CacheServer) StartLeaderHeartbeatMonitor() {
+	// wait for decision to get leader
+	s.logger.Info("Leader heartbeat monitor starting...")
+
+	ticker := time.NewTicker(time.Second)
+	for {
+		// run heartbeat check every 1 second
+		<-ticker.C
+
+		// case 1: we are a follower
+		if s.leaderID != s.nodeID {
+			if !s.IsLeaderAlive() {
+				s.logger.Info("Leader heartbeat failed, running new election")
+				s.RunElection()
+				s.logger.Info("Election done, leader heartbeat continuing")
+			}
+
+			select {
+			case <-s.shutdownChan:
+				s.logger.Info("Received shutdown signal")
+				break
+			case <-time.After(time.Second):
+				continue
+			}
+
+			// case 2: we are the leader, so check for any dead nodes and remove them from cluster
+		} else {
+			modified := false
+			for _, node := range s.nodesConfig.Nodes {
+				// skip self
+				if node.Id == s.nodeID {
+					continue
+				}
+				// new identity service client
+				c, err := s.NewCacheClient(node.Host, int(node.Port))
+				if err != nil {
+					s.logger.Infof("error creating grpc client to node node %s: %v", node.Id, err)
+					delete(s.nodesConfig.Nodes, node.Id)
+					modified = true
+					continue
+				}
+
+				ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+				defer cancel()
+
+				s.logger.Infof("Checking health of node %s", node.Id)
+				_, err = c.GetHeartbeat(ctx, &HeartbeatRequest{CallerNodeId: s.nodeID})
+				if err != nil {
+					s.logger.Infof("Node %s healthcheck returned error, removing from cluster", node.Id)
+					delete(s.nodesConfig.Nodes, node.Id)
+					modified = true
+				}
+			}
+
+			// if cluster was modified, send out updated cluster config to other nodes
+			if modified {
+				s.logger.Info("Detected node config change, sending update to other nodes")
+				s.updateClusterConfigInternal()
+			}
+		}
+	}
+}
+
+// Check if leader node is alive (3 second timeout)
+func (s *CacheServer) IsLeaderAlive() bool {
+	// make sure leader exists
+	if s.leaderID == NO_LEADER {
+		s.logger.Infof("IsLeaderAlive found leader doesn't exist")
+		return false
+	}
+	// if this node is the leader, return true
+	if s.nodeID == s.leaderID {
+		return true
+	}
+	s.logger.Infof("leader is %s", s.leaderID)
+	leader, ok := s.nodesConfig.Nodes[s.leaderID]
+	if !ok {
+		s.logger.Infof("leader %s does not exist", s.leaderID)
+		return true
+	}
+
+	// new identity service client
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+
+	// make status request rpc
+	c, err := s.NewCacheClient(leader.Host, int(leader.Port))
```

Comment:
recreation of gRPC connection is cost a lot, don't recreate it every time

pb all ypur `Node` structs already have a `gclient`,  just remember leader node in configuration and take connection based on this info